wallet/wallet.go

```diff
@@ -3655,70 +3691,24 @@ func (w *Wallet) publishTransaction(tx *wire.MsgTx) (*chainhash.Hash, error) {
 		return nil, err
 	}
 
-	// match is a helper method to easily string match on the error
-	// message.
-	match := func(err error, s string) bool {
-		return strings.Contains(strings.ToLower(err.Error()), s)
-	}
-
-	_, err = chainClient.SendRawTransaction(tx, false)
-
-	// Determine if this was an RPC error thrown due to the transaction
-	// already confirming.
-	var rpcTxConfirmed bool
-	if rpcErr, ok := err.(*btcjson.RPCError); ok {
-		rpcTxConfirmed = rpcErr.Code == btcjson.ErrRPCTxAlreadyInChain
-	}
-
 	var (
 		txid      = tx.TxHash()
 		returnErr error
 	)
 
-	switch {
-	case err == nil:
+	_, err = chainClient.SendRawTransaction(tx, false)
+	if err == nil {
 		return &txid, nil
+	}
 
-	// Since we have different backends that can be used with the wallet,
-	// we'll need to check specific errors for each one.
-	//
-	// If the transaction is already in the mempool, we can just return now.
-	//
-	// This error is returned when broadcasting/sending a transaction to a
-	// btcd node that already has it in their mempool.
-	// https://github.com/btcsuite/btcd/blob/130ea5bddde33df32b06a1cdb42a6316eb73cff5/mempool/mempool.go#L953
-	case match(err, "already have transaction"):
-		fallthrough
+	returnErr = MapBroadcastBackendError(err)
 
-	// This error is returned when broadcasting a transaction to a bitcoind
-	// node that already has it in their mempool.
-	// https://github.com/bitcoin/bitcoin/blob/9bf5768dd628b3a7c30dd42b5ed477a92c4d3540/src/validation.cpp#L590
-	case match(err, "txn-already-in-mempool"):
+	switch returnErr.(type) {
```

Comment:
not sure, in this case it would work because the error mapping does only return an empty error type without including the backend error for those two examples, but this could easily change so I think we would need to use errors.As instead which brings a bit of overhead because we would need to declare the variables as well because errors.As does only accept a non-nil target, so I would keep it as is ? Or should we always try to use errors.As instead of the switch err.(type) statement?

Reply:
> Or should we always try to use errors.As instead of the switch err.(type) statement?

I think the "modern Go" best practice is to prefer `errors.As` as it can handle wrapped errors, while type assertions can only go one level deep.

Reply:
This would look something like this with `errors.As`:
```go
returnErr = MapBroadcastBackendError(err)

var errInMempool *ErrInMempool
var errAlreadyConfirmed *ErrAlreadyConfirmed

switch {
case errors.As(returnErr, &errInMempool):
	log.Infof("%v: tx already in mempool", txid)
	return &txid, nil

case errors.As(returnErr, &errAlreadyConfirmed):
	dbErr := walletdb.Update(w.db, func(dbTx walletdb.ReadWriteTx) error {
		txmgrNs := dbTx.ReadWriteBucket(wtxmgrNamespaceKey)
		txRec, err := wtxmgr.NewTxRecordFromMsgTx(tx, time.Now())
		if err != nil {
			return err
		}
		return w.TxStore.RemoveUnminedTx(txmgrNs, txRec)
	})
	if dbErr != nil {
		log.Warnf("Unable to remove confirmed transaction %v "+
			"from unconfirmed store: %v", tx.TxHash(), dbErr)
	}

	log.Infof("%v: tx already confirmed", txid)

	return &txid, nil
}
```